crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -2,797 +2,535 @@
-        if !is_final {
-            trace!(
-                ?effects_digest,
-                "tx cert/effects not yet final: {} < {}",
-                *cur_stake,
-                quorum_threshold
-            );
-        }
-        is_final
-    }
-
-    pub fn forget_effects(&mut self, digests: &TransactionEffectsDigest) {
-        self.effects_stake_map.remove(digests);
-        self.effects_vote_map.remove(digests);
-    }
-}
+async fn follower_process<A, Handler>(
+    node_sync_handle: Handler,
+    node_sync_store: Arc<NodeSyncStore>,
+    aggregator: Arc<AuthorityAggregator<A>>,
+    max_stream_items: u64,
+    metrics: &GossipMetrics,
+    mut cancel_receiver: oneshot::Receiver<()>,
+) where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+    Handler: SyncHandler + Clone,
+{
+    debug!("starting follower process");
 
-/// Waiter is used to single-shot concurrent requests and wait for dependencies to finish.
-struct Waiter<Key, ResultT> {
-    waiters: Mutex<HashMap<Key, broadcast::Sender<ResultT>>>,
-}
+    let mut backoff = HashMap::new();
 
-impl<Key, ResultT> Waiter<Key, ResultT>
-where
-    Key: std::hash::Hash + Eq + Clone + std::fmt::Debug,
-    ResultT: Clone,
-{
-    fn new() -> Self {
-        Self {
-            waiters: Mutex::new(HashMap::new()),
-        }
-    }
+    loop {
+        let mut follower_tasks = FuturesUnordered::new();
+        let mut reconnects = FuturesUnordered::new();
 
-    /// Returns (true, rx) if there are no other waiters yet, or else (false, rx).
-    /// All rxes can be woken by calling notify(key, result)
-    async fn wait(&self, key: &Key) -> (bool, broadcast::Receiver<ResultT>) {
-        let waiters = &mut self.waiters.lock().unwrap();
-        let entry = waiters.entry(key.clone());
-
-        match entry {
-            hash_map::Entry::Occupied(e) => (false, e.get().subscribe()),
-            hash_map::Entry::Vacant(e) => {
-                let (tx, rx) = broadcast::channel(1);
-                e.insert(tx);
-                (true, rx)
+        let start_one_task = |name, handle, store| {
+            let start_time = Instant::now();
+            let client = aggregator.clone_client(name);
+            async move {
+                let result =
+                    follow_one_peer(handle, store, *name, client, max_stream_items, metrics)
+                        .await
+                        .tap_err(|e| warn!("follower task exited with error {}", e));
+                (result, start_time, name)
             }
-        }
-    }
+        };
 
-    async fn notify(&self, key: &Key, res: ResultT) -> SuiResult {
-        if let Some(tx) = self.waiters.lock().unwrap().remove(key) {
-            tx.send(res).map_err(|_| SuiError::GenericAuthorityError {
-                error: format!("couldn't notify waiters for key {:?}", key),
-            })?;
-        } else {
-            trace!("no pending waiters");
+        for (name, _) in aggregator.committee.members() {
```

Comment:
just to double note that in the old logic, one FN follows `|committee| - 1` validators but here it is changed. There are pros and cons (cons especially when the committee size is large) and we may need to revisit it in the future

Reply:
The old code subtracts one because it assumes that it is a validator, and it doesn't want to follow itself.

Fullnodes should absolutely follow the entire committee - they require finality, and the only way to be sure that you will discover finality is to follow the entire committee. (You need 2f+1 to attest to the cert. Any subset of the committee may contain up to f byzantine nodes. Therefore there is no proper subset that is guaranteed to contain 2f+1 honest nodes).

Reply:
following all validators to get quorum assumes the worst case, which we should certainly prepare for. But in majority cases, we will not have f faulty nodes. Even if we have a handful (<f), this can be fixed by connecting with different nodes in the next round (similarly to how do we it in the current code where we randomly choose many but not all of the validators). 

The trade-off is load on validators, maybe we need to worry about it after this PR, or checkpoint fully works, but sth to keep in mind

Reply:
The solution to the scalability problems is to not have every fullnode in the world connect directly to the validators. We shouldn't compromise liveness here in order to put a bandaid on a larger problem.